src/ui/pages/ConnectedSites/ConnectedSites.tsx

```diff
@@ -143,38 +149,131 @@ function ConnectedSitesList({
+            <VStack gap={4}>
+              <div>No connected DApps found</div>
+              {hasFilters ? (
+                <UnstyledButton
+                  onClick={onReset}
+                  style={{ color: 'var(--primary)' }}
+                  className={helperStyles.hoverUnderline}
+                >
+                  Reset all filters
+                </UnstyledButton>
+              ) : null}
+            </VStack>
+          </UIText>
+        </VStack>
+      ) : (
+        <UIText
+          kind="body/regular"
+          color="var(--neutral-500)"
+          style={{ padding: 20, textAlign: 'center' }}
+        >
+          You will see a list of connected DApps here
+        </UIText>
+      )}
+    </FillView>
+  );
+}
+
 function ConnectedSitesMain() {
-  const { data: connectedSites, ...connectedSitesQuery } = useQuery({
+  const {
+    data: allConnectedSites,
+    isLoading,
+    ...connectedSitesQuery
+  } = useQuery({
     queryKey: ['getPermissionsWithWallets'],
     queryFn: getPermissionsWithWallets,
     useErrorBoundary: true,
     suspense: true,
   });
+  const [searchQuery, setSearchQuery] = useState<string | undefined>();
+  const searchInputRef = useRef<InputHandle | null>(null);
 
-  if (connectedSitesQuery.isLoading) {
-    return <p>Loading qweasd...</p>;
-  }
+  const filteredConnectedSites = useMemo(() => {
+    if (!searchQuery) {
+      return allConnectedSites;
+    }
+    const query = searchQuery.trim();
+    return allConnectedSites?.filter((site) =>
+      site.origin.toLowerCase().includes(query)
+    );
+  }, [allConnectedSites, searchQuery]);
 
   return (
     <PageColumn>
-      {connectedSites?.length ? (
+      <PageTop />
+      {allConnectedSites?.length ? (
         <>
-          <Spacer height={16} />
-          <ConnectedSitesList
-            items={connectedSites}
-            onRevokeAll={() => connectedSitesQuery.refetch()}
+          <ConnectedSitesSearch
+            inputRef={searchInputRef}
+            value={searchQuery}
+            onChange={setSearchQuery}
           />
+          <Spacer height={24} />
         </>
+      ) : null}
+      {isLoading ? null : filteredConnectedSites?.length ? (
+        <ConnectedSitesList
+          items={filteredConnectedSites}
+          onRevokeAll={() => connectedSitesQuery.refetch()}
+        />
       ) : (
-        <FillView>
-          <UIText
-            kind="body/regular"
-            color="var(--neutral-500)"
-            style={{ padding: 20, textAlign: 'center' }}
-          >
-            You will see a list of connected DApps here
-          </UIText>
-        </FillView>
+        <EmptyView
+          hasConnectedSites={Boolean(allConnectedSites?.length)}
+          hasFilters={Boolean(searchQuery)}
+          onReset={() => {
+            if (searchInputRef.current) {
+              searchInputRef.current.setValue('');
+            }
+            setSearchQuery(undefined);
```

Comment:
Trying to follow this practice and avoid persisting "undefined" values: https://x.com/sebmarkbage/status/1774082542372520441